osu.Game/Graphics/Containers/SectionsContainer.cs

```diff
@@ -195,16 +204,49 @@ protected override void UpdateAfterChildren()
                 headerBackgroundContainer.Y = ExpandableHeader?.Y ?? 0;
 
                 float scrollOffset = FixedHeader?.LayoutSize.Y ?? 0;
-                Func<T, float> diff = section => scrollContainer.GetChildPosInContent(section) - currentScroll - scrollOffset;
 
-                if (scrollContainer.IsScrolledToEnd())
+                float visibleHeight(T section)
+                {
+                    float sectionPosition = scrollContainer.GetChildPosInContent(section) - scrollOffset;
+                    float top = Math.Max(sectionPosition, currentScroll);
+                    float bottom = Math.Min(sectionPosition + section.Height, currentScroll + scrollContainer.DisplayableContent);
+                    return bottom - top;
+                }
+
+                var sortedByVisible = new List<T>(Children);
+                sortedByVisible.Sort((a, b) => visibleHeight(b).CompareTo(visibleHeight(a)));
+                T mostVisible = sortedByVisible.FirstOrDefault();
+
+                if (mostVisible == null)
+                    return;
+
+                int mostVisibleIndex = IndexOf(mostVisible);
+
+                bool usePreviousSection()
                 {
-                    SelectedSection.Value = Children.LastOrDefault();
+                    if (mostVisibleIndex == 0)
+                        return false;
+
+                    T sectionBefore = Children[mostVisibleIndex - 1];
+                    // consider the section user wants to see smaller than it actually is to lower the visibility requirement
+                    // this makes the section get selected sooner when scrolling upwards, to match it already getting selected sooner when scrolling downwards
+                    float sectionHeight = sectionBefore.Height * (lastClickedSection as T == sectionBefore ? 0.8f : 1.0f);
+
+                    bool moreVisible = Precision.AlmostBigger(visibleHeight(sectionBefore), visibleHeight(Children[mostVisibleIndex]), 1);
+                    bool completelyVisible = Precision.AlmostBigger(visibleHeight(sectionBefore), sectionHeight, 1);
+
+                    return moreVisible || completelyVisible;
                 }
-                else
+
+                // we want to use 100% visible previous sections instead because otherwise small sections sandwiched between large section would never get selected
+                while (usePreviousSection())
```

Comment:
i think the function name here is misleading/confusing. "while use previous section is true" means nothing to me if i read it out loud.